src/binary-reader-objdump.cc

```diff
@@ -32,77 +32,247 @@ namespace {
 
 typedef std::vector<uint32_t> Uint32Vector;
 
-struct Context {
-  ObjdumpOptions* options;
-  Stream* out_stream;
-  const uint8_t* data;
-  size_t size;
-  Opcode current_opcode;
-  size_t current_opcode_offset;
-  size_t last_opcode_end;
-  int indent_level;
-  bool print_details;
-  bool header_printed;
-  int section_found;
-
+class BinaryReaderObjdumpPrepass : public BinaryReaderNop {
+ public:
+  BinaryReaderObjdumpPrepass(const uint8_t* data,
+                             size_t size,
+                             ObjdumpOptions* options);
+
+  virtual Result OnFunctionName(uint32_t function_index,
+                                StringSlice function_name);
+  virtual Result OnRelocCount(uint32_t count,
+                              BinarySection section_code,
+                              StringSlice section_name);
+  virtual Result OnReloc(RelocType type,
+                         uint32_t offset,
+                         uint32_t index,
+                         int32_t addend);
+
+ protected:
+  bool ShouldPrintDetails();
+  void PrintDetails(const char* fmt, ...);
+
+  ObjdumpOptions* options = nullptr;
+  Stream* out_stream = nullptr;
+  const uint8_t* data = nullptr;
+  size_t size = 0;
+  bool print_details = false;
+  BinarySection reloc_section = BinarySection::Invalid;
   uint32_t section_starts[kBinarySectionCount];
-  BinarySection reloc_section;
-
-  uint32_t next_reloc;
 };
 
-}  // namespace
+BinaryReaderObjdumpPrepass::BinaryReaderObjdumpPrepass(const uint8_t* data,
+                                                       size_t size,
+                                                       ObjdumpOptions* options)
+    : options(options), data(data), size(size) {
+  WABT_ZERO_MEMORY(section_starts);
+  out_stream = init_stdout_stream();
+}
 
-static bool should_print_details(Context* ctx) {
-  if (ctx->options->mode != ObjdumpMode::Details)
+bool BinaryReaderObjdumpPrepass::ShouldPrintDetails() {
+  if (options->mode != ObjdumpMode::Details)
     return false;
-  return ctx->print_details;
+  return print_details;
 }
 
-static void WABT_PRINTF_FORMAT(2, 3)
-    print_details(Context* ctx, const char* fmt, ...) {
-  if (!should_print_details(ctx))
+void WABT_PRINTF_FORMAT(2, 3)
+    BinaryReaderObjdumpPrepass::PrintDetails(const char* fmt, ...) {
+  if (!ShouldPrintDetails())
     return;
   va_list args;
   va_start(args, fmt);
   vprintf(fmt, args);
   va_end(args);
 }
 
-static Result begin_section(BinaryReaderContext* ctx,
-                            BinarySection section_code,
-                            uint32_t size) {
-  Context* context = static_cast<Context*>(ctx->user_data);
-  context->section_starts[static_cast<size_t>(section_code)] = ctx->offset;
+Result BinaryReaderObjdumpPrepass::OnFunctionName(uint32_t index,
+                                                  StringSlice name) {
+  if (options->mode == ObjdumpMode::Prepass) {
```

Comment:
Why is this condition needed?  Shouldn't it just be split into two different methods?  On a related note is seems strange for BinaryReaderObjdump to inherit from BinaryReaderObjdumpPrepass.  Could they both inherit from some common base class instead?